flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/RowDataToProtoSerializer.java

```diff
@@ -0,0 +1,256 @@
+                            "Expected Type: %s at Field Number %d for Logical Type: %s.\nError: %s",
+                            fieldDescriptor.getType().name(),
+                            fieldNumber,
+                            fieldType.getTypeRoot().name(),
+                            e.getMessage()));
+        }
+        return null;
+    }
+
+    /**
+     * Method to extract Array Element Type from ARRAY datatype.
+     *
+     * @param fieldType The ARRAY Datatype.
+     * @return ARRAY element type.
+     * @throws UnsupportedOperationException in case ARRAY has NULL datatype or multiple datatypes.
+     */
+    private static LogicalType getArrayElementType(LogicalType fieldType)
+            throws UnsupportedOperationException {
+        List<LogicalType> arrayElementTypes = fieldType.getChildren();
+        if (fieldType.isNullable()) {
+            throw new UnsupportedOperationException("NULLABLE ARRAY is not supported.");
+        }
+        if (arrayElementTypes.size() > 1) {
+            throw new UnsupportedOperationException(
+                    "Multiple Datatypes not supported in ARRAY type");
+        }
+        LogicalType arrayElementType = arrayElementTypes.get(0);
+        if (arrayElementType.getTypeRoot() == LogicalTypeRoot.NULL) {
+            throw new UnsupportedOperationException("ARRAY of type NULL is not supported.");
+        }
+        return arrayElementType;
+    }
+
+    private Object convertArrayElement(
+            ArrayData.ElementGetter elementGetter,
+            RowData element,
+            int fieldNumber,
+            int pos,
+            LogicalType arrayElementType,
+            FieldDescriptor fieldDescriptor) {
+        Object ele = elementGetter.getElementOrNull(element.getArray(fieldNumber), pos);
+        return toProtoValue(arrayElementType, 0, GenericRowData.of(ele), fieldDescriptor);
+    }
+
+    /**
+     * Function to convert a Generic Avro Record to Dynamic Message to write using the Storage Write
+     * API.
+     *
+     * @param element {@link GenericRecord} Object to convert to {@link DynamicMessage}
+     * @param descriptor {@link Descriptor} describing the schema of the sink table.
+     * @return {@link DynamicMessage} Object converted from the Generic Avro Record.
+     */
+    public DynamicMessage getDynamicMessageFromRowData(
+            RowData element, Descriptor descriptor, LogicalType type) {
+        DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+        // Get a record's field schema and find the field descriptor for each field one by
+        // one.
+        int fieldNumber = 0;
+        for (LogicalType fieldType : type.getChildren()) {
+            // In case no field descriptor exists for the field, throw an error as we have
+            // incompatible schemas.
+            Descriptors.FieldDescriptor fieldDescriptor =
+                    Preconditions.checkNotNull(descriptor.findFieldByNumber(fieldNumber + 1));
```

Comment:
can you explain to me the +1 here?

Reply:
Field Numbers are 0 onwards while descriptor fields are added from 1.

Reply:
thanks! you can resolve this convo